doc/man/man1/ldappasswd.1
.TH LDAPPASSWD 1 "RELEASEDATE" "OpenLDAP LDVERSION"
.\" $OpenLDAP$
.\" Copyright 1998-2021 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.SH NAME
ldappasswd \- change the password of an LDAP entry
.SH SYNOPSIS
.B ldappasswd
[\c
.BR \-V [ V ]]
[\c
.BI \-d \ debuglevel\fR]
[\c
.BR \-n ]
[\c
.BR \-v ]
[\c
.BR \-A ]
[\c
.BI \-a \ oldPasswd\fR]
[\c
.BI \-t \ oldpasswdfile\fR]
[\c
.BR \-S ]
[\c
.BI \-s \ newPasswd\fR]
[\c
.BI \-T \ newpasswdfile\fR]
[\c
.BR \-x ]
[\c
.BI \-D \ binddn\fR]
[\c
.BR \-W ]
[\c
.BI \-w \ passwd\fR]
[\c
.BI \-y \ passwdfile\fR]
[\c
.BI \-H \ ldapuri\fR]
[\c
.BI \-h \ ldaphost\fR]
[\c
.BI \-p \ ldapport\fR]
[\c
.BR \-e \ [ ! ] \fIext\fP [ =\fIextparam\fP ]]
[\c
.BR \-E \ [ ! ] \fIext\fP [ =\fIextparam\fP ]]
[\c
.BI \-o \ opt \fR[= optparam \fR]]
[\c
.BI \-O \ security-properties\fR]
[\c
.BR \-I ]
[\c
.BR \-Q ]
[\c
.BR \-N ]
[\c
.BI \-U \ authcid\fR]
[\c
.BI \-R \ realm\fR]
[\c
.BI \-X \ authzid\fR]
[\c
.BI \-Y \ mech\fR]
[\c
.BR \-Z [ Z ]]
[\c
.IR user ]
.SH DESCRIPTION
.B ldappasswd
is a tool to set the password of an LDAP user.
.B ldappasswd
uses the LDAPv3 Password Modify (RFC 3062) extended operation.
.LP
.B ldappasswd
sets the password of associated with the user [or an optionally
specified
.IR user ]. 
If the new
password is not specified on the command line and the user
doesn't enable prompting, the server will be asked to generate
a password for the user.
.LP
.B ldappasswd
is neither designed nor intended to be a replacement for
.BR passwd (1)
and should not be installed as such.
.SH OPTIONS
.TP
.BR \-V [ V ]
Print version info.
If \fB\-VV\fP is given, only the version information is printed.
.TP
.BI \-d \ debuglevel
Set the LDAP debugging level to \fIdebuglevel\fP.
.B ldappasswd
must be compiled with LDAP_DEBUG defined for this option to have any effect.
.TP
.B \-n
Do not set password. (Can be useful when used in conjunction with
\fB\-v\fP or \fB\-d\fP)
.TP
.B \-v
Increase the verbosity of output.  Can be specified multiple times.
.TP
.BI \-A
Prompt for old password.
This is used instead of specifying the password on the command line.
.TP
.BI \-a \ oldPasswd
Set the old password to \fIoldPasswd\fP.
.TP
.BI \-t \ oldPasswdFile
Set the old password to the contents of \fIoldPasswdFile\fP.
.TP
.BI \-S
Prompt for new password.
This is used instead of specifying the password on the command line.
.TP
.BI \-s \ newPasswd
Set the new password to \fInewPasswd\fP.
.TP
.BI \-T \ newPasswdFile
Set the new password to the contents of \fInewPasswdFile\fP.
.TP
.B \-x 
Use simple authentication instead of SASL.
.TP
.BI \-D \ binddn
Use the Distinguished Name \fIbinddn\fP to bind to the LDAP directory.
For SASL binds, the server is expected to ignore this value.
.TP
.BI \-W
Prompt for bind password.
This is used instead of specifying the password on the command line.
.TP
.BI \-w \ passwd
Use \fIpasswd\fP as the password to bind with.
.TP
.BI \-y \ passwdfile
Use complete contents of \fIpasswdfile\fP as the password for
simple authentication.
.TP
.BI \-H \ ldapuri
Specify URI(s) referring to the ldap server(s); only the protocol/host/port
fields are allowed; a list of URI, separated by whitespace or commas
is expected.
.TP
.BI \-h \ ldaphost
Specify an alternate host on which the ldap server is running.
Deprecated in favor of \fB\-H\fP.
.TP
.BI \-p \ ldapport
Specify an alternate TCP port where the ldap server is listening.
Deprecated in favor of \fB\-H\fP.
.TP
.BR \-e \ [ ! ] \fIext\fP [ =\fIextparam\fP ]
.TP
.BR \-E \ [ ! ] \fIext\fP [ =\fIextparam\fP ]

Specify general extensions with \fB\-e\fP and passwd modify extensions with \fB\-E\fP.
\'\fB!\fP\' indicates criticality.

General extensions:
.nf
  [!]assert=<filter>    (an RFC 4515 Filter)
  !authzid=<authzid>    ("dn:<dn>" or "u:<user>")
  [!]bauthzid           (RFC 3829 authzid control)
  [!]chaining[=<resolve>[/<cont>]]
  [!]manageDSAit
  [!]noop
  ppolicy
  [!]postread[=<attrs>] (a comma-separated attribute list)
  [!]preread[=<attrs>]  (a comma-separated attribute list)
  [!]relax
  sessiontracking
  abandon,cancel,ignore (SIGINT sends abandon/cancel,
  or ignores response; if critical, doesn't wait for SIGINT.
  not really controls)
.fi

Passwd Modify extensions:
.nf
  (none)
.fi
.TP
.BI \-o \ opt \fR[= optparam \fR]]

Specify general options.

General options:
.nf
  nettimeout=<timeout>  (in seconds, or "none" or "max")
  ldif-wrap=<width>     (in columns, or "no" for no wrapping)
.fi
.TP
.BI \-O \ security-properties
Specify SASL security properties.
.TP
.B \-I
Enable SASL Interactive mode.  Always prompt.  Default is to prompt
only as needed.
.TP
.B \-Q
Enable SASL Quiet mode.  Never prompt.
.TP
.B \-N
Do not use reverse DNS to canonicalize SASL host name.
.TP
.BI \-U \ authcid
Specify the authentication ID for SASL bind. The form of the ID
depends on the actual SASL mechanism used.
.TP
.BI \-R \ realm
Specify the realm of authentication ID for SASL bind. The form of the realm
depends on the actual SASL mechanism used.
.TP
.BI \-X \ authzid
Specify the requested authorization ID for SASL bind.
.I authzid
must be one of the following formats:
.BI dn: "<distinguished name>"
or
.BI u: <username>\fP.
.TP
.BI \-Y \ mech
Specify the SASL mechanism to be used for authentication. If it's not
specified, the program will choose the best mechanism the server knows.
.TP
.BR \-Z [ Z ]
Issue StartTLS (Transport Layer Security) extended operation. If you use
\fB\-ZZ\fP, the command will require the operation to be successful
.SH SEE ALSO
.BR ldap_sasl_bind (3),
.BR ldap_extended_operation (3),
.BR ldap_start_tls_s (3)
.SH AUTHOR
The OpenLDAP Project <http://www.openldap.org/>
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.4 Change Log

OpenLDAP 2.4.57 Release (2021/01/18)
	Fixed ldapexop to use correct return code (ITS#9417)
	Fixed slapd to remove asserts in UUIDNormalize (ITS#9391)
	Fixed slapd to remove assert in csnValidate (ITS#9410)
	Fixed slapd validity checks for issuerAndThisUpdateCheck (ITS#9411, ITS#9427)
	Fixed slapd validity checks for serialNumberAndIssuerCheck (ITS#9404, ITS#9424)
	Fixed slapd AVA sort with invalid RDN (ITS#9412)
	Fixed slapd ldap_X509dn2bv to check for invalid BER after RDN count (ITS#9423, ITS#9425)
	Fixed slapd saslauthz to remove asserts in validation (ITS#9406, ITS#9407)
	Fixed slapd saslauthz to use slap_sl_free on normalized DN (ITS#9409)
	Fixed slapd saslauthz SEGV in slap_parse_user (ITS#9413)
	Fixed slapd modrdn memory leak (ITS#9420)
	Fixed slapd double-free in vrfilter (ITS#9408)
	Fixed slapd cancel operation to correctly terminate (ITS#9428)
	Fixed slapd-ldap fix binds on retry with closed connection (ITS#9400)
	Fixed slapo-syncprov to ignore duplicate sessionlog entries (ITS#9394)

OpenLDAP 2.4.56 Release (2020/11/10)
	Fixed slapd to remove assert in certificateListValidate (ITS#9383)
	Fixed slapd to remove assert in csnNormalize23 (ITS#9384)
	Fixed slapd to better parse ldapi listener URIs (ITS#9379)

OpenLDAP 2.4.55 Release (2020/10/26)
	Fixed slapd normalization handling with modrdn (ITS#9370)
	Fixed slapd-meta to check ldap_install_tls return code (ITS#9366)
	Contrib
		Fixed nssov misplaced semicolon (ITS#8731, ITS#9368)

OpenLDAP 2.4.54 Release (2020/10/12)
	Fixed slapd delta-syncrepl to ignore delete ops on deleted entry (ITS#9342)
	Fixed slapd delta-syncrepl to be fully serialized (ITS#9330)
	Fixed slapd delta-syncrepl MOD on zero-length context entry (ITS#9352)
	Fixed slapd syncrepl to be fully serialized (ITS#8102)
	Fixed slapd syncrepl to call check_syncprov on fresh consumer (ITS#9345)
	Fixed slapd syncrepl to propagate errors from overlay_entry_get_ov (ITS#9355)
	Fixed slapd syncrepl to not create empty ADD ops (ITS#9359)
	Fixed slapd syncrepl replace usage on single valued attrs (ITS#9295)
	Fixed slapd-monitor fix monitor_back_register_database for empty suffix DB (ITS#9353)
	Fixed slapo-accesslog normalizer for reqStart (ITS#9358)
	Fixed slapo-accesslog to not generate new contextCSN on purge (ITS#9361)
	Fixed slapo-syncprov contextCSN generation with empty suffix (ITS#9015)
	Fixed slapo-syncprov sessionlog to use a TAVL tree (ITS#8486)

OpenLDAP 2.4.53 Release (2020/09/07)
	Added slapd syncrepl additional SYNC logging (ITS#9043)
	Fixed slapd syncrepl segfault on NULL cookie on REFRESH (ITS#9282)
	Fixed slapd syncrepl to use fresh connection on REFRESH fallback (ITS#9338)
	Fixed slapo-ppolicy race condition for pwdFailureTime (ITS#9302,ITS#9334)
	Build
		Require OpenSSL 1.0.2 or later (ITS#9323)
		Fixed libldap compilation issue with broken C compilers (ITS#9332)

OpenLDAP 2.4.52 Release (2020/08/28)
	Added libldap LDAP_OPT_X_TLS_REQUIRE_SAN option (ITS#9318)
	Added libldap OpenSSL support for multiple EECDH curves (ITS#9054)
	Added slapd OpenSSL support for multiple EECDH curves (ITS#9054)
	Fixed librewrite malloc/free corruption (ITS#9249)
	Fixed libldap hang when using UDP and server down (ITS#9328)
	Fixed slapd syncrepl rare deadlock due to network issues (ITS#9324)
	Fixed slapd syncrepl regression that could trigger an assert (ITS#9329)
	Fixed slapd-mdb index error with collapsed range (ITS#9135)

OpenLDAP 2.4.51 Release (2020/08/11)
	Added slapo-ppolicy implement Netscape password policy controls (ITS#9279)
	Fixed libldap retry loop in ldap_int_tls_connect (ITS#8650)
	Fixed libldap to use getaddrinfo in ldap_pvt_get_fqdn (ITS#9287)
	Fixed slapd to enforce singular existence of some overlays (ITS#9309)
	Fixed slapd syncrepl to not delete non-replicated attrs (ITS#9227)
	Fixed slapd syncrepl to correctly delete entries on resync (ITS#9282)
	Fixed slapd syncrepl to use replace on single valued attrs (ITS#9294, ITS#9295)
	Fixed slapd-perl dynamic config with threaded slapd (ITS#7573)
	Fixed slapo-ppolicy to expose the ppolicy control (ITS#9285)
	Fixed slapo-ppolicy race condition for pwdFailureTime (ITS#9302)
	Fixed slapo-ppolicy so it can only exist once per DB (ITS#9309)
	Fixed slapo-chain to check referral (ITS#9262)
	Build Environment
		Fix test064 so it no longer uses bashisms (ITS#9263)
	Contrib
		Fix default prefix value for pw-argon2, pw-pbkdf2 modules (ITS#9248)
		slapo-allowed - Fix usage of unitialized variable (ITS#9308)
	Documentation
		ldap_parse_result(3) - Document ldap_parse_intermediate (ITS#9271)

OpenLDAP 2.4.50 Release (2020/04/28)
	Fixed client benign typos (ITS#8890)
	Fixed libldap type cast (ITS#9175)
	Fixed libldap retry loop in ldap_int_tls_connect (ITS#8650)
	Fixed libldap_r race on Windows mutex initialization (ITS#9181)
	Fixed liblunicode memory leak (ITS#9198)
	Fixed slapd benign typos (ITS#8890)
	Fixed slapd to limit depth of nested filters (ITS#9202)
	Fixed slapd-mdb memory leak in dnSuperiorMatch (ITS#9214)
	Fixed slapo-pcache database initialization (ITS#9182)
	Fixed slapo-ppolicy callback (ITS#9171)
	Build
		Fix olcDatabaseDummy initialization for windows (ITS#7074)
		Fix detection for ws2tcpip.h for windows (ITS#8383)
		Fix back-mdb types for windows (ITS#7878)
	Contrib
		Update ldapc++ config.guess and config.sub to support newer architectures (ITS#7855)
		Added pw-argon2 module (ITS#9233, ITS#8575, ITS#9203, ITS#9206)
	Documentation
		slapd-ldap(5) - Clarify idassert-authzfrom behavior (ITS#9003)
		slapd-meta(5) - Remove client-pr option (ITS#8683)
		slapindex(8) - Fix truncate option information for back-mdb (ITS#9230)

OpenLDAP 2.4.49 Release (2020/01/30)
	Added slapd-monitor database entry count for slapd-mdb (ITS#9154)
	Fixed client tools to not add controls on cancel/abandon (ITS#9145)
	Fixed client tools SyncInfo message to be LDIF compliant (ITS#8116)
	Fixed libldap to correctly free sb (ITS#9081, ITS#8755)
	Fixed libldap descriptor leak if ldaps fails (ITS#9147)

clients/tools/ldappasswd.c
/* ldappasswd -- a tool for change LDAP passwords */
/* $OpenLDAP$ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 1998-2021 The OpenLDAP Foundation.
 * Portions Copyright 1998-2003 Kurt D. Zeilenga.
 * Portions Copyright 1998-2001 Net Boolean Incorporated.
 * Portions Copyright 2001-2003 IBM Corporation.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in the file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* Portions Copyright (c) 1992-1996 Regents of the University of Michigan.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that this notice is preserved and that due credit is given
 * to the University of Michigan at Ann Arbor.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.  This
 * software is provided ``as is'' without express or implied warranty.
 */
/* ACKNOWLEDGEMENTS:
 * The original ldappasswd(1) tool was developed by Dave Storey (F5
 * Network), based on other OpenLDAP client tools (which are, of
 * course, based on U-MICH LDAP).  This version was rewritten
 * by Kurt D. Zeilenga (based on other OpenLDAP client tools).
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/socket.h>
#include <ac/string.h>
#include <ac/time.h>
#include <ac/unistd.h>

#include <ldap.h>
#include "lutil.h"
#include "lutil_ldap.h"
#include "ldap_defaults.h"

#include "common.h"


static struct berval newpw = { 0, NULL };
static struct berval oldpw = { 0, NULL };

static int   want_newpw = 0;
static int   want_oldpw = 0;

static char *oldpwfile = NULL;
static char *newpwfile = NULL;

void
usage( void )
{
	fprintf( stderr, _("Change password of an LDAP user\n\n"));
	fprintf( stderr,_("usage: %s [options] [user]\n"), prog);
	fprintf( stderr, _("  user: the authentication identity, commonly a DN\n"));
	fprintf( stderr, _("Password change options:\n"));
	fprintf( stderr, _("  -a secret  old password\n"));
	fprintf( stderr, _("  -A         prompt for old password\n"));
	fprintf( stderr, _("  -t file    read file for old password\n"));
	fprintf( stderr, _("  -s secret  new password\n"));
	fprintf( stderr, _("  -S         prompt for new password\n"));
	fprintf( stderr, _("  -T file    read file for new password\n"));
	tool_common_usage();
	exit( EXIT_FAILURE );
}


const char options[] = "a:As:St:T:"
	"d:D:e:h:H:InNO:o:p:QR:U:vVw:WxX:y:Y:Z";

int
handle_private_option( int i )
{
	switch ( i ) {
#if 0
	case 'E': /* passwd extensions */ {
		int		crit;
		char	*control, *cvalue;
		if( protocol == LDAP_VERSION2 ) {
			fprintf( stderr, _("%s: -E incompatible with LDAPv%d\n"),
			         prog, protocol );
			exit( EXIT_FAILURE );
		}

		/* should be extended to support comma separated list of
		 *	[!]key[=value] parameters, e.g.  -E !foo,bar=567
		 */

		crit = 0;
		cvalue = NULL;
		if( optarg[0] == '!' ) {
			crit = 1;
			optarg++;
		}

		control = strdup( optarg );
		if ( (cvalue = strchr( control, '=' )) != NULL ) {
			*cvalue++ = '\0';
		}
		fprintf( stderr, _("Invalid passwd extension name: %s\n"), control );
		usage();
		}
#endif

	case 'a':	/* old password (secret) */
		oldpw.bv_val = strdup( optarg );
		{
			char* p;
			for( p = optarg; *p != '\0'; p++ ) {
				*p = '\0';
			}
		}
		oldpw.bv_len = strlen( oldpw.bv_val );
		break;

	case 'A':	/* prompt for old password */
		want_oldpw++;
		break;

	case 's':	/* new password (secret) */
		newpw.bv_val = strdup (optarg);
		{
			char* p;
			for( p = optarg; *p != '\0'; p++ ) {
				*p = '\0';
			}
		}
		newpw.bv_len = strlen( newpw.bv_val );
		break;

	case 'S':	/* prompt for user password */
		want_newpw++;
		break;

	case 't':
		oldpwfile = optarg;
		break;

	case 'T':
		newpwfile = optarg;
		break;

	default:
		return 0;
	}
	return 1;
}


int
main( int argc, char *argv[] )
{
	int rc;
	char	*user = NULL;

	LDAP	       *ld = NULL;
	struct berval bv = {0, NULL};
	BerElement  *ber = NULL;

	int id, code = LDAP_OTHER;
	LDAPMessage *res;
	char *matcheddn = NULL, *text = NULL, **refs = NULL;
	char	*retoid = NULL;
	struct berval *retdata = NULL;
	LDAPControl **ctrls = NULL;

    tool_init( TOOL_PASSWD );
	prog = lutil_progname( "ldappasswd", argc, argv );

	/* LDAPv3 only */
	protocol = LDAP_VERSION3;

	tool_args( argc, argv );

	if( argc - optind > 1 ) {
		usage();
	} else if ( argc - optind == 1 ) {
		user = strdup( argv[optind] );
	} else {
		user = NULL;
	}

	if( oldpwfile ) {
		rc = lutil_get_filed_password( oldpwfile, &oldpw );
		if( rc ) {
			rc = EXIT_FAILURE;
			goto done;
		}
	}

	if( want_oldpw && oldpw.bv_val == NULL ) {
		/* prompt for old password */
		char *ckoldpw;
		oldpw.bv_val = strdup(getpassphrase(_("Old password: ")));
		ckoldpw = getpassphrase(_("Re-enter old password: "));

		if( oldpw.bv_val == NULL || ckoldpw == NULL ||
			strcmp( oldpw.bv_val, ckoldpw ))
		{
			fprintf( stderr, _("passwords do not match\n") );
			rc = EXIT_FAILURE;
			goto done;
		}

		oldpw.bv_len = strlen( oldpw.bv_val );
	}

	if( newpwfile ) {
		rc = lutil_get_filed_password( newpwfile, &newpw );
		if( rc ) {
			rc = EXIT_FAILURE;
			goto done;
		}
	}

	if( want_newpw && newpw.bv_val == NULL ) {
		/* prompt for new password */
		char *cknewpw;
		newpw.bv_val = strdup(getpassphrase(_("New password: ")));
		cknewpw = getpassphrase(_("Re-enter new password: "));

		if( newpw.bv_val == NULL || cknewpw == NULL ||
			strcmp( newpw.bv_val, cknewpw ))
		{
			fprintf( stderr, _("passwords do not match\n") );
			rc = EXIT_FAILURE;
			goto done;
		}

		newpw.bv_len = strlen( newpw.bv_val );
	}

	ld = tool_conn_setup( 0, 0 );

	tool_bind( ld );

	if( user != NULL || oldpw.bv_val != NULL || newpw.bv_val != NULL ) {
		/* build the password modify request data */
		ber = ber_alloc_t( LBER_USE_DER );

		if( ber == NULL ) {
			perror( "ber_alloc_t" );
			rc = EXIT_FAILURE;
			goto done;
		}

		ber_printf( ber, "{" /*}*/ );

		if( user != NULL ) {
			ber_printf( ber, "ts",
				LDAP_TAG_EXOP_MODIFY_PASSWD_ID, user );
			free(user);
		}

		if( oldpw.bv_val != NULL ) {
			ber_printf( ber, "tO",
				LDAP_TAG_EXOP_MODIFY_PASSWD_OLD, &oldpw );
			free(oldpw.bv_val);
		}

		if( newpw.bv_val != NULL ) {
			ber_printf( ber, "tO",
				LDAP_TAG_EXOP_MODIFY_PASSWD_NEW, &newpw );
			free(newpw.bv_val);
		}

		ber_printf( ber, /*{*/ "N}" );

		rc = ber_flatten2( ber, &bv, 0 );

		if( rc < 0 ) {
			perror( "ber_flatten2" );
			rc = EXIT_FAILURE;
			goto done;
		}
	}

	if ( dont ) {
		rc = LDAP_SUCCESS;
		goto done;
	}

	tool_server_controls( ld, NULL, 0);

	rc = ldap_extended_operation( ld,
		LDAP_EXOP_MODIFY_PASSWD, bv.bv_val ? &bv : NULL, 
		NULL, NULL, &id );

	ber_free( ber, 1 );

	if( rc != LDAP_SUCCESS ) {
		tool_perror( "ldap_extended_operation", rc, NULL, NULL, NULL, NULL );
		rc = EXIT_FAILURE;
		goto done;
	}

	for ( ; ; ) {
		struct timeval	tv;

		if ( tool_check_abandon( ld, id ) ) {
			tool_exit( ld, LDAP_CANCELLED );
		}

		tv.tv_sec = 0;
		tv.tv_usec = 100000;

		rc = ldap_result( ld, LDAP_RES_ANY, LDAP_MSG_ALL, &tv, &res );
		if ( rc < 0 ) {
			tool_perror( "ldap_result", rc, NULL, NULL, NULL, NULL );
			tool_exit( ld, rc );
		}

		if ( rc != 0 ) {
			break;
		}
	}

	rc = ldap_parse_result( ld, res,
		&code, &matcheddn, &text, &refs, &ctrls, 0 );
	if( rc != LDAP_SUCCESS ) {
		tool_perror( "ldap_parse_result", rc, NULL, NULL, NULL, NULL );
		rc = EXIT_FAILURE;
		goto done;
	}

	rc = ldap_parse_extended_result( ld, res, &retoid, &retdata, 1 );
	if( rc != LDAP_SUCCESS ) {
		tool_perror( "ldap_parse_extended_result", rc, NULL, NULL, NULL, NULL );
		rc = EXIT_FAILURE;
		goto done;
	}

	if( retdata != NULL ) {
		ber_tag_t tag;
		char *s;
		ber = ber_init( retdata );

		if( ber == NULL ) {
			perror( "ber_init" );
			rc = EXIT_FAILURE;
			goto done;
		}

		/* we should check the tag */
		tag = ber_scanf( ber, "{a}", &s);

		if( tag == LBER_ERROR ) {
			perror( "ber_scanf" );
		} else {
			printf(_("New password: %s\n"), s);
			ber_memfree( s );
		}

		ber_free( ber, 1 );

	} else if ( code == LDAP_SUCCESS && newpw.bv_val == NULL ) {
		tool_perror( "ldap_parse_extended_result", LDAP_DECODING_ERROR,
			" new password expected", NULL, NULL, NULL );
	}

	if( verbose || code != LDAP_SUCCESS ||
		( matcheddn && *matcheddn ) || ( text && *text ) || refs || ctrls )
	{
		printf( _("Result: %s (%d)\n"), ldap_err2string( code ), code );

		if( text && *text ) {
			printf( _("Additional info: %s\n"), text );
		}

		if( matcheddn && *matcheddn ) {
			printf( _("Matched DN: %s\n"), matcheddn );
		}

		if( refs ) {
			int i;
			for( i=0; refs[i]; i++ ) {
				printf(_("Referral: %s\n"), refs[i] );
			}
		}

		if( ctrls ) {
			tool_print_ctrls( ld, ctrls );
			ldap_controls_free( ctrls );
		}
	}

	ber_memfree( text );
	ber_memfree( matcheddn );
	ber_memvfree( (void **) refs );
	ber_memfree( retoid );
	ber_bvfree( retdata );

	rc = ( code == LDAP_SUCCESS ) ? EXIT_SUCCESS : EXIT_FAILURE;

done:
	/* disconnect from server */
	tool_exit( ld, rc ); 
}
